Update I18N design pattern content (#244)
pxblue/doc-it: src/docs/patterns/internationalization.mdx

import UIFlippedImage from '../../app/assets/design-patterns/internationalization/ui-flipped.png';
import ButtonWidthImage from '../../app/assets/design-patterns/internationalization/button-width.png';
import TextFieldWidthImage from '../../app/assets/design-patterns/internationalization/text-field-width.png';
import LanguageOptionsImage from '../../app/assets/design-patterns/internationalization/language-options.png';
import CulturalDifferencesImage from '../../app/assets/design-patterns/internationalization/cultural-differences.svg';

<!-- KEYWORDS: design pattern internationalization i18n internationalisation language translate translation localization l10n rtl localisation globalization globalize -->

# Internationalization & Localization

<ImageGrid images={[InternationalizationImage]} regularWidth />
<DemoCard
repository={'internationalization'}
angular={
'https://stackblitz.com/github/pxblue/angular-design-patterns?file=src%2Fapp%2Fpages%2Fi18n%2Fi18n.component.ts'
}
react={
'https://codesandbox.io/s/github/pxblue/react-design-patterns/tree/master?file=/src/pages/i18n/index.tsx&initialpath=i18n'
}
/>

> Internationalization and localization referes to a sequence of practices in adapting applications to various regions, languages, and cultures. You can accomplish this in your applications by considering the differences in language, scripts, and directionality.

Reviewer: ```suggestion
Internationalization and localization refer to a set of practices for adapting applications to different regions, languages, and cultures. You can accomplish this in your applications by considering the differences in language, scripts, and directionality.
```

> Internationalization, sometimes abreviated as "I18N", refers to the process of designing and developing a generic base product, free of systematic biases in language and culture. Localization, sometimes written as "L10N", talks about adapting or customizing a product to some specific areas. Internationalization and localization are both crucial in making your application prosper in a global market.

Reviewer: Great paragraph

<Divider />

# Changing Locales

<MaterialDesignDescription
    title={'Language Support'}
    url={'https://material.io/design/typography/language-support.html#language-considerations'}
/>

## Translating the Text

To localize your application content, you should create a string resource file or object that contains the translations for any display text used in your application. You will not want to have any text hard-coded in your application's UI. To switch your application to another language, you simply toggle the file/object used to populate the UI to the appropriate translation.

## Showing Language Options

<ImageGrid images={[LanguageOptionsImage]} regularWidth />

> Language options are typically offered in the user preference settings. In case when a user authorization is absent (such as an application available to everyone on-site, or a public-facing website), the language option is expected in the app bar or in the footer.

Reviewer: ```suggestion
Language options are typically presented in the user preferences / settings section of an application. In the case where users do not require an account (such as a public-facing website), the language option is expected in the app bar or in the footer.
```

When showing a list of languages, you can decide whether you want to translate the language names or not (e.g., "English" vs. "Inglés"). There are pros and cons to each approach, so consider what makes the most sense for your application. Providing other cues (such as an icon) to help users find the language selection can help.

## Converting Units, Currencies, and Formats

> While this usually goes with the locale settings, certain users might have different preferences, and there can be variations even within the same locale. You should provide the user with default settings when they first load a locale, but allow the users change their preferences, and remember it the next time they log in with that locale.

Reviewer: ```suggestion
Units, currency, and date / time formats are usually linked with a particular locale. However, there can be differences in usage even within a single locale. You should provide users with the option to change these values (if appropriate for your application) in their locale settings.
```

<MaterialDesignDescription
    title={'Bidirectionality'}
    description={'Learn about mirroring layout, mirroring elements, and localization.'}
    url={'https://material.io/design/usability/bidirectionality.html'}
/>

Some languages, such as Arabic and Hebrew, are read from right to left. Applications supporting these languages will need to mirror the UI so the natural flow is from right to left. For example, navigation drawers should be displayed on the right side of the screen, icons should be to the right of the text in buttons, and any icons with directionality should be mirrored.

<ImageGrid
    images={[UIFlippedImage]}
    regularWidth
    caption={[
        '1: A PX Blue user interface in right-to-left language (English).',
        '2: The same PX Blue user interface in left-to-right language (Arabic).',
    ]}
/>
<ImageGrid
    images={[IconsFlippedImage]}
    regularWidth
    caption={[
        '1: Icons that indicate directions should be mirrored.',
        '2: Icons with glyphs and symbols should never be mirrored.',
    ]}
/>

## Cultural Differences

<ImageGrid
    images={[CulturalDifferencesImage]}
    regularWidth

>     caption={'People from different cultures are appealed to different things.'}

Reviewer: ```suggestion
```

/>

> Addressing the cultural differences is the key concern of the localization. Sometimes certain UI elements convey different meaning in different cultures. For example, in America, the red color in the stock market indicates a decline, while in China the same color indicates a rise. A thumb-up gesture usually expresses an approval in English-speaking country, yet the gesture can have a negative meaning in Iraq. When designing for a new market, you are encouraged to have a local consultant run an audit for your application.

Reviewer: ```suggestion
Addressing cultural differences is the key concern of localization. Sometimes certain UI elements convey different meaning in different cultures. For example, in America, the red color in the stock market indicates a decline, while in China the same color indicates a rise. A thumbs-up gesture usually expresses an approval in English-speaking country, yet the gesture can have a negative meaning in Iraq. When designing for a new market, you are encouraged to have a local consultant run an audit for your application.
```

When you are designing for webpages (instead of applications), your primary goal is to advertise your products and services, and appeal to the local market. The language tone, page style, and choice of image assets are expected to be tailored for different targeted culture. Likewise, your user study and market research conclusions derived under a certain cultural background might not apply to another culture.

<React.Fragment>
    <MaterialDesignDescription
        title={'Design Localization and the New Web'}
        description={'An examination of localization in web design by Bitmatica Lab.'}
        url={'https://medium.com/bitmatica-lab/design-localization-and-the-new-web-f54f43c5f040#'}

>         avatar={<Web style={{ width: 48, height: 48 }} />}

Reviewer: Not a huge fan of this Web icon for these.
What about 
![image](https://user-images.githubusercontent.com/29152776/92471629-3ee0b380-f1a6-11ea-8226-421c2c4d1ea4.png)
or
![image](https://user-images.githubusercontent.com/29152776/92471659-4acc7580-f1a6-11ea-9006-85db4b3b04e9.png)

Author: I am using the link icon right now... feels a bit weird

        description={`Apple's suggestions on designing for an international market.`}
        avatar={<Web style={{ width: 48, height: 48 }} />}
    />
    <MaterialDesignDescription
        title={'Localization - Market Finder'}
        description={'Introduction to localization by Google Market Finder.'}
        url={'https://marketfinder.thinkwithgoogle.com/intl/en_us/topic/localisation#'}
        avatar={<Web style={{ width: 48, height: 48 }} />}
    />
    <Divider />
</React.Fragment>

# Developers

There are a number of available libraries to assist with implementing internationalization in your applications.

## Angular and Ionic

For projects that need the ability to swap between languages, we recommend using [ngx-translate](https://github.com/ngx-translate/core). As an alternative, the Angular framework has a powerful built-in [i18n tool](https://angular.io/guide/i18n) available, but it requires a separate web app be built for each supported language (e.g. an English version, a Spanish version, etc). This is ideal for larger projects that don't require in-app language switching and don't support more than a few locales.

## React

> There are many JS libraries available to help internationalize your applications. We recommend using a third-party solution [react-i18next](https://github.com/i18next/react-i18next).

Reviewer: ```suggestion
There are many JS libraries available to help internationalize your applications. We recommend using [react-i18next](https://github.com/i18next/react-i18next) to manage your translations. If you are supporting right-to-left languages, you should also carefully follow the guide from [Material-UI](https://material-ui.com/guides/right-to-left).
```

## React Native

> For mirroring the user interface, we recommend using [I18nManager](https://reactnative.dev/blog/2016/08/19/right-to-left-support-for-react-native-apps). To manage your translation string, we recommend [react-i18next](https://react.i18next.com/).

Reviewer: ```suggestion
As with React for web, we recommend using [react-i18next](https://github.com/i18next/react-i18next) to manage your translations. If you are supporting right-to-left languages, you can make use of the [I18nManager](https://reactnative.dev/blog/2016/08/19/right-to-left-support-for-react-native-apps) from React Native.
```

Internationalization is the practice of adapting applications to various regions, languages, and cultures. You can accomplish this in your applications by considering the differences in language, scripts, and directionality.

<Divider />

# Steps for Converting from One Locale to Another Locale

<React.Fragment>
    <MaterialDesignDescription
        title={'Language Support'}
        url={'https://material.io/design/typography/language-support.html#language-considerations'}
    />
    <MaterialDesignDescription
        title={'Localization'}
        url={'https://developer.apple.com/localization/'}
        description={`Apple's suggestions on designing for an international market.`}
        avatar={<Apple style={{ width: 48, height: 48 }} />}
    />
</React.Fragment>

## Showing Language Options

When showing a list of languages, you can decide whether you want to translate the language names or not (e.g., "English" vs. "Inglés"). There are pros and cons to each approach, so consider what makes the most sense for your application. Providing other cues (such as an icon) to help users find the language selection can help.

## Translating the Text

> To localize your application content, you should create a string resource file or object that contains the translations for any display text used in your application. You will not want to have any text hard-coded in your application's UI. To switch your application to another language, you simply toggle the file/object used to populate the UI to the appropriate translation.

Author: localization = address the **_cultural_** difference, such as the tone, the images, etc, more about marketing (so usually applicable to physical product design and web design, but less about application design)

internationalization = everything else, so that you can render different languages, units, currencies... without breaking your UI

## Converting Units, Currencies, and Number Formats

While this usually goes with the locale settings, certain users might have different preferences, and there can be variations even within the same locale. You should provide the user with default settings when they first load a locale, but allow the users change their preferences, and remember it the next time they login with that locale.

## Checking for Spacing

Different languages/writing systems may require different vertical and horizontal space. To ensure flexibility, avoid using fixed dimensions for necessary UI components such as buttons, labels, and badges. Also, consider setting different default display font sizes for languages that may have smaller glyphs.

> **NOTE:** PX Blue recommends using [Noto Sans](https://www.google.com/get/noto/#sans-lgc) for all non-western languages. You can read more on our [Typography](/style/typography) page.

## Bi-directionality/Right-to-Left (RTL) Support

<MaterialDesignDescription
    title={'Bidirectionality'}
    description={'Learn about mirroring layout, mirroring elements, and localization.'}
    url={'https://material.io/design/usability/bidirectionality.html'}
/>

> Some languages, such as Arabic and Hebrew, are read from right to left. Applications supporting these languages will need to mirror the UI so the natural flow is from right to left. For example, navigation drawers should be displayed on the right side of the screen, icons should be to the right of the text in buttons, and any icons with directionality should be mirrored.

Reviewer: Should we also mention the flipping of certain icons (and link to the wonderful deep link where Material explains which to flip)?

Author: That wonderful link is right on the top, between line 56 and line 60.

Reviewer: Yeah, but the part about which icons to flip is 4 miles down on that Material.io page. It's probably fine, I was just wondering if it might be nice to deep link specifically to the icon part in this context.

Author: Added an additional inline deeplink

<ImageGrid
    images={[UIFlippedImage]}
    regularWidth
    caption={[
        '1: A PX Blue user interface in right-to-left language (English).',
        '2: The same PX Blue user interface in left-to-right language (Arabic).',
    ]}
/>
<ImageGrid
    images={[IconsFlippedImage]}
    regularWidth
    caption={[
        '1: Icons that indicate directions should be mirrored.',
        '2: Icons with glyphs and symbols should never be mirrored.',
    ]}
/>

> ## Curtural differences

Reviewer: ```suggestion
## Cultural Differences
```

Author: Should have thrown things in spell checkers 🤦‍♀️

Sometimes certain UI elements convey different meaning in different cultures. For example, in America, the red color in the stock market indicates a decline, while in China the same color indicates a rise. A thumb-up gesture usually expresses an approval in English-speaking country, yet the gesture can have a negative meaning in Iraq. When designing for a new market, you are encouraged to have a local consultant run an audit for your application.

<Divider />

# Developers

There are a number of available libraries to assist with implementing internationalization in your applications.

> ## Angular

Reviewer: and Ionic?

Author: I wasn’t sure what library to use for Ionic. Should I just change the headline to “Angular and Ionic”?

Reviewer: I'm assuming it would use the same as angular since it's just a wrapper....similar to using all the same components

Reviewer: We haven't actually done an RTL/I18N example in Ionic yet though

For projects that need the ability to swap between languages, we recommend using [ngx-translate](https://github.com/ngx-translate/core). As an alternative, the Angular framework has a powerful built-in [i18n tool](https://angular.io/guide/i18n) available, but it requires a separate web app be built for each supported language (e.g. an English version, a Spanish version, etc). This is ideal for larger projects that don't require in-app language switching and don't support more than a few locales.

> ## React and React Native

Reviewer: Material-UI is not applicable to React Native.

Reviewer: You'll want to mention the I18nManager from React Native (for dealing with RTL) and also i18next/react-i18next for translations.

> There are many JS libraries available to help internationalize your application. If you are already using PX Blue, [Material UI](https://material-ui.com/guides/localization/) has a lightweight solution available.

Reviewer: Material-UIs solution is very very limited (I wouldn't even call it lightweight). It basically just does some minor localization of things like calendars and currencies if I'm not mistaken. I would push the i18next as the primary library to use.

Author: No mentioning of https://material-ui.com/guides/right-to-left/?

Reviewer: Oh yeah, definitely mention that. That part is _really_ good. The localization guide is meh.

<ImageGrid images={[LanguageOptionsImage]} regularWidth />

Language options are typically presented in the user preferences / settings section of an application. In the case where users do not require an account (such as a public-facing website), the language option is expected in the app bar or in the footer. You should provide cues (such as an icon) to help users locate the language selection.

> When showing a list of languages, the language name themselves should not be translated (e.g., use "Español" instead of "Spanish").

Reviewer: ```suggestion
When showing a list of languages, the language names should be presented in their own language (i.e., use "Español", not "Spanish").
```
